Sources/LeafErrorMiddleware/LeafErrorMiddleware.swift

```diff
@@ -1,61 +1,79 @@
 import Vapor
-import HTTP
+import Leaf
 
-public final class LeafErrorMiddleware: Middleware {
-    let log: LogProtocol
+/// Captures all errors and transforms them into an internal server error.
+public final class LeafErrorMiddleware: Middleware, Service {
+    /// The environment to respect when presenting errors.
     let environment: Environment
-    let viewRenderer: ViewRenderer
-    public init(environment: Environment, log: LogProtocol, viewRenderer: ViewRenderer) {
-        self.log = log
+
+    /// Log destination
+    let log: Logger
+
+    /// Create a new ErrorMiddleware for the supplied environment.
+    public init(environment: Environment, log: Logger) {
         self.environment = environment
-        self.viewRenderer = viewRenderer
+        self.log = log
     }
 
-    public func respond(to req: Request, chainingTo next: Responder) throws -> Response {
-        do {
-            return try next.respond(to: req)
-        } catch {
-            log.error(error)
-            return make(with: req, for: error)
-        }
-    }
+    /// See `Middleware.respond`
+    public func respond(to req: Request, chainingTo next: Responder) throws -> Future<Response> {
+        let renderer = try req.make(LeafRenderer.self)
+        let promise = req.eventLoop.newPromise(Response.self)
+
+        func handleError(with status: HTTPStatus) {
+            if status == .notFound {
+                do {
+                    try renderer
+                        .render("404")
+                        .encode(for: req)
+                        .do { res in
+                            res.http.status = status
+                            promise.succeed(result: res)
+                    }
+                } catch { /* swallow so we return the default view */ }
+            }
 
-    public func make(with req: Request, for error: Error) -> Response {
-        let status: Status = Status(error)
-        if status == .notFound {
```

Comment:
Extra case for authorization errors ?

Reply:
There are two templates that the middleware renders - one for 404 errors and one for 500 errors (or everything else) - that's why the check is here